src/nerdbank-streams/src/Utilities.ts

```diff
@@ -1,7 +1,58 @@
 import CancellationToken from "cancellationtoken";
+import { Readable, Writable } from "stream";
 import { Deferred } from "./Deferred";
 import { IDisposableObservable } from "./IDisposableObservable";
 
+export async function writeAsync(stream: NodeJS.WritableStream, chunk: Buffer) {
+    const deferred = new Deferred<void>();
+    stream.write(chunk, (err?: Error) => {
+        if (err) {
+            deferred.reject(err);
+        } else {
+            deferred.resolve();
+        }
+    });
+    return deferred.promise;
+}
+
+export function writeSubstream(stream: NodeJS.WritableStream): NodeJS.WritableStream {
+    return new Writable({
+        async write(chunk: Buffer, encoding: string, callback: (error?: Error | null) => void) {
+            try {
+                const dv = new DataView(new ArrayBuffer(4));
+                dv.setUint32(0, chunk.length, false);
+                await writeAsync(stream, Buffer.from(dv.buffer));
+                await writeAsync(stream, chunk);
+                callback();
+            } catch (err) {
+                callback(err);
+            }
+        },
+        final(callback: (error?: Error | null) => void) {
+            // Write the terminating 0 length sequence.
+            stream.write(new Uint8Array(4), callback);
+        },
+    });
+}
+
+export function readSubstream(stream: NodeJS.ReadableStream): NodeJS.ReadableStream {
+    return new Readable({
+        async read(size: number) {
+            const lenBuffer = await getBufferFrom(stream, 4);
+            const dv = new DataView(lenBuffer.buffer, lenBuffer.byteOffset, lenBuffer.length);
+            const chunkSize = dv.getUint32(0, false);
+            if (chunkSize === 0) {
+                this.push(null);
+                return;
+            }
+
+            // TODO: make this *stream* instead of read as an atomic chunk.
```

Comment:
This isn't a blocker for the basic functionality that you're looking for, @jepetty. I'm working on addressing this now, but it may be in a subsequent PR.

Reply:
soon today.